src/geocel/VolumeCollectionBuilder.hh

```diff
@@ -0,0 +1,94 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file geocel/VolumeCollectionBuilder.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <type_traits>
+
+#include "corecel/Types.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/data/CollectionBuilder.hh"
+
+#include "Types.hh"
+
+namespace celeritas
+{
+class GeoParamsInterface;
+//---------------------------------------------------------------------------//
+/*!
+ * Create a "collection" that maps from ImplVolumeId to a value.
+ *
+ * This builds collections for runtime execution using a GeoTrackView's \c
+ * impl_volume_id() call, which requires less indirection than \c volume_id() .
+ *
+ * Given a geometry (which is allowed to be \c GeoParamsInterface ), a function
+ * \code T(*)(VolumeId) \endcode will be called for every \c ImplVolumeId in
+ * the geometry that corresponds to a canonical volume. The resulting value
+ * will be assigned to the collection.
+ */
+template<class T, class G, class F>
+inline Collection<T, Ownership::value, MemSpace::host, ImplVolumeId>
+build_volume_collection(G const& geo, F&& fill_value)
+{
+    static_assert(std::is_base_of_v<GeoParamsInterface, G>,
+                  "G must be a geometry class");
+    static_assert(std::is_convertible_v<std::invoke_result_t<F, VolumeId>, T>,
+                  "Incorrect return type for F");
+
+    Collection<T, Ownership::value, MemSpace::host, ImplVolumeId> result;
+    auto num_impl_volumes = geo.impl_volumes().size();
+    resize(&result, num_impl_volumes);
+    for (auto iv_id : range(ImplVolumeId{num_impl_volumes}))
+    {
+        T value{};
+        if (auto vol_id = geo.volume_id(iv_id))
+        {
+            value = fill_value(vol_id);
+        }
+        result[iv_id] = value;
```

Comment:
Is `T` guaranteed to be trivial/fast to copy?  Otherwise consider:
```suggestion
        if (auto vol_id = geo.volume_id(iv_id))
        {
            result[iv_id] = fill_value(vol_id);
        } else {
	        result[iv_id] = T{};
	    }
```
or if `Collection` has `emplace`:
```suggestion
        if (auto vol_id = geo.volume_id(iv_id))
        {
            result.emplace(std::make_pair(iv_id, fill_value(vol_id)));
        } else {
	        result.emplace(std::make_pair(iv_id, T{}));
	    }
```

Reply:
I've rearranged it to improve clarity and eliminate the unnecessary assignment.